src/scheduler/controller.rs

```diff
@@ -0,0 +1,102 @@
+
+pub struct GCController<VM: VMBinding> {
+    mmtk: &'static MMTK<VM>,
+    requester: Arc<ControllerCollectorContext<VM>>,
+    scheduler: Arc<GCWorkScheduler<VM>>,
+    receiver: Receiver<CoordinatorMessage<VM>>,
+    coordinator_worker: Box<GCWorker<VM>>,
+}
+
+impl<VM: VMBinding> GCController<VM> {
+    pub fn new(
+        mmtk: &'static MMTK<VM>,
+        requester: Arc<ControllerCollectorContext<VM>>,
+        scheduler: Arc<GCWorkScheduler<VM>>,
+        receiver: Receiver<CoordinatorMessage<VM>>,
+        coordinator_worker: Box<GCWorker<VM>>,
+    ) -> Box<GCController<VM>> {
+        Box::new(Self {
+            mmtk,
+            requester,
+            scheduler,
+            receiver,
+            coordinator_worker,
+        })
+    }
+
+    pub fn run(&mut self, tls: VMWorkerThread) {
+        // Initialize the GC worker for coordinator. We are not using the run() method from
+        // GCWorker so we manually initialize the worker here.
+        self.coordinator_worker.tls = tls;
+
+        loop {
+            debug!("[STWController: Waiting for request...]");
+            self.requester.wait_for_request();
+            debug!("[STWController: Request recieved.]");
+
+            // For heap growth logic
+            // FIXME: This is not used. However, we probably want to set a 'user_triggered' flag
+            // when GC is requested.
+            // let user_triggered_collection: bool = SelectedPlan::is_user_triggered_collection();
+
+            self.scheduler.set_initializer(Some(ScheduleCollection));
+            self.wait_for_completion();
+            debug!("[STWController: Worker threads complete!]");
+        }
+    }
+
+    /// Drain the message queue and execute coordinator work.
+    pub fn wait_for_completion(&mut self) {
+        let worker = &mut self.coordinator_worker;
+        let mmtk = self.mmtk;
+
+        // At the start of a GC, we probably already have received a `ScheduleCollection` work. Run it now.
+        if let Some(mut initializer) = self.scheduler.take_initializer() {
+            initializer.do_work_with_stat(worker, mmtk);
+        }
+        loop {
+            let message = self.receiver.recv().unwrap();
+            match message {
+                CoordinatorMessage::Work(mut work) => {
+                    work.do_work_with_stat(worker, mmtk);
+                }
+                CoordinatorMessage::AllWorkerParked | CoordinatorMessage::BucketDrained => {
+                    self.scheduler.update_buckets();
+                }
+            }
+            let _guard = self.scheduler.worker_monitor.0.lock().unwrap();
+            if self.scheduler.worker_group().all_parked() && self.scheduler.all_buckets_empty() {
+                break;
+            }
+        }
+        for message in self.receiver.try_iter() {
+            if let CoordinatorMessage::Work(mut work) = message {
+                work.do_work_with_stat(worker, mmtk);
+                //self.process_coordinator_work(work);
```

Comment:
This line could be removed. Alternatively, you can implement this method in this struct and use it instead of `work.do_work_with_stat()`. It may be helpful if we want to log stuff for coordinator work.

Reply:
I'll use `work.do_work_with_stata(worker, mmtk);`.  Using `self.` will borrow `self`, and Rust seems to have problem borrowing `self` mutably and `self.receiver` immutably at the same time.